File: internal/parse/parse.go
// Package parse provides shared parsing utilities for sizes, seed times, and numbers.
// These are used by the stats fetching layer and the history recorder.
package parse

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var sizeRe = regexp.MustCompile(`(?i)([\d.]+)\s*(B|[KMGTP]i?B)`)

var sizeFactors = map[string]float64{
	"b":   1.0 / (1024 * 1024 * 1024),
	"kib": 1.0 / (1024 * 1024),
	"mib": 1.0 / 1024,
	"gib": 1.0,
	"tib": 1024.0,
	"pib": 1024.0 * 1024.0,
	// Decimal-labelled units (TBDev-family sites render "1.005 TB"): tracker
	// software near-universally computes 1024-based sizes but mislabels them,
	// so these map to the same factors as their -iB counterparts.
	"kb": 1.0 / (1024 * 1024),
	"mb": 1.0 / 1024,
	"gb": 1.0,
	"tb": 1024.0,
	"pb": 1024.0 * 1024.0,
}

// SizeToGiB parses a human-readable size string (e.g. "3.14 TiB") into GiB.
// Returns nil when the input is empty or unparseable.
func SizeToGiB(s string) *float64 {
	m := sizeRe.FindStringSubmatch(s)
	if m == nil {
		return nil
	}
	v, err := strconv.ParseFloat(m[1], 64)
	if err != nil {
		return nil
	}
	f, ok := sizeFactors[strings.ToLower(m[2])]
	if !ok {
		return nil
	}
	result := v * f
	return &result
}

// SeedTimeToSeconds parses a seed time string (e.g. "3M 6D 22h 23m 13s") to seconds.
// Returns nil for empty/invalid input.
// Also accepts a plain integer/float as raw seconds.
func SeedTimeToSeconds(s string) *float64 {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}
	// Plain numeric — treat as seconds directly
	if v, err := strconv.ParseFloat(s, 64); err == nil {
		if v > 0 {
			return &v
		}
		return nil
	}
	tokens := durTokenRe.FindAllStringSubmatch(s, -1)
	if len(tokens) == 0 {
		return nil
	}
	units := make([]durUnit, len(tokens))
	for i, t := range tokens {
		units[i] = classifyDurUnit(t[2])
	}
	// A bare lowercase "m" means minutes in English layouts and months in
	// Portuguese ones ("2m 4d 13h 59m 13s" is 2 months … 59 minutes). UNIT3D
	// renders units in descending order, so an "m" with a day, week or hour
	// after it occupies the month slot; every later one is minutes.
	for i, t := range tokens {
		if units[i] != durUnknown || t[2] != "m" {
			continue
		}
		units[i] = durMinute
		for _, later := range units[i+1:] {
			if later == durDay || later == durWeek || later == durHour {
				units[i] = durMonth
				break
			}
		}
	}
	var total float64
	for i, t := range tokens {
		v, err := strconv.ParseFloat(t[1], 64)
		if err != nil {
			continue
		}
		total += v * durSeconds[units[i]]
	}
	if total <= 0 {
		return nil
	}
	return &total
}

// durTokenRe splits a duration into value/unit pairs, e.g. "178a 2m 4d 20h
// 52m 38s" → (178,a) (2,m) (4,d) (20,h) (52,m) (38,s). Every occurrence of a
// unit counts, so a string carrying both a month and a minute "m" adds both.
var durTokenRe = regexp.MustCompile(`([\d.]+)\s*([A-Za-zÀ-ÿ]+)`)

type durUnit int

const (
	durUnknown durUnit = iota
	durYear
	durMonth
	durWeek
	durDay
	durHour
	durMinute
	durSecond
)

var durSeconds = map[durUnit]float64{
	durUnknown: 0,
	durYear:    365.25 * 86400,
	durMonth:   30.44 * 86400,
	durWeek:    7 * 86400,
	durDay:     86400,
	durHour:    3600,
	durMinute:  60,
	durSecond:  1,
}

// classifyDurUnit maps a unit token to its duration unit. Case matters where
// tracker layouts rely on it: "M" is months against "m" minutes in English,
// and "S" is semanas (weeks) against "s" seconds on Portuguese sites. A bare
// lowercase "m" is ambiguous and gets resolved by position in the caller.
func classifyDurUnit(raw string) durUnit {
	switch raw {
	case "y", "Y", "a", "A", "ano", "anos", "year", "years":
		return durYear
	case "M", "mo", "mes", "meses", "month", "months":
		return durMonth
	case "w", "W", "S", "sem", "semana", "semanas", "week", "weeks":
		return durWeek
	case "d", "D", "dia", "dias", "day", "days":
		return durDay
	case "h", "H", "hora", "horas", "hour", "hours":
		return durHour
	case "min", "mins", "minuto", "minutos", "minute", "minutes":
		return durMinute
	case "s", "seg", "segs", "segundo", "segundos", "sec", "secs", "second", "seconds":
		return durSecond
	}
	return durUnknown
}

// AnyFloat converts any numeric JSON value to float64.
func AnyFloat(v any) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case int:
		return float64(n)
	case string:
		f, _ := strconv.ParseFloat(strings.ReplaceAll(n, ",", ""), 64)
		return f
	}
	return 0
}

// BytesToSize converts a raw byte count (as returned by the Gazelle API) to a
// human-readable size string with exactly 2 decimal places, e.g. "25.52 GiB".
func BytesToSize(b int64) string {
	if b <= 0 {
		return "0.00 B"
	}
	const (
		kib = int64(1024)
		mib = kib * 1024
		gib = mib * 1024
		tib = gib * 1024
		pib = tib * 1024
	)
	switch {
	case b >= pib:
		return fmt.Sprintf("%.2f PiB", float64(b)/float64(pib))
	case b >= tib:
		return fmt.Sprintf("%.2f TiB", float64(b)/float64(tib))
	case b >= gib:
		return fmt.Sprintf("%.2f GiB", float64(b)/float64(gib))
	case b >= mib:
		return fmt.Sprintf("%.2f MiB", float64(b)/float64(mib))
	case b >= kib:
		return fmt.Sprintf("%.2f KiB", float64(b)/float64(kib))
	default:
		return fmt.Sprintf("%d B", b)
	}
}

// NormalizeSeedSize re-formats a size string to exactly 2 decimal places.
// Handles scraped values like "1.239 TiB" → "1.24 TiB".
// Leaves strings that don't match a known unit pattern unchanged.
func NormalizeSeedSize(s string) string {
	m := sizeRe.FindStringSubmatch(s)
	if m == nil {
		return s
	}
	v, err := strconv.ParseFloat(m[1], 64)
	if err != nil {
		return s
	}
	return fmt.Sprintf("%.2f %s", v, m[2])
}

// FormatSeedTime formats seconds into a human string like "1Y 2M 3D 4h 5m 6s".
func FormatSeedTime(totalSec float64) string {
	t := int64(totalSec)
	if t <= 0 {
		return "0s"
	}
	steps := []struct {
		sec int64
		u   string
	}{
		{31536000, "Y"}, {2592000, "M"}, {604800, "W"},
		{86400, "D"}, {3600, "h"}, {60, "m"}, {1, "s"},
	}
	var parts []string
	for _, s := range steps {
		v := t / s.sec
		t -= v * s.sec
		if v > 0 {
			parts = append(parts, fmt.Sprintf("%d%s", v, s.u))
		}
	}
	if len(parts) == 0 {
		return "0s"
	}
	return strings.Join(parts, " ")
}

File: web/src/utils/parse.ts
// utils/parse.ts — parsing helpers for sizes, times, dates

/** Parse a size string (e.g. "3.14 TiB") → GiB as number, or null */
export function parseSize(str: unknown): number | null {
  if (!str) return null;
  const m = String(str).match(/([\d.]+)\s*(B|[KMGTP]i?B)/i);
  if (!m) return null;
  // Decimal-labelled units (kb/mb/gb/tb) map to the binary factors — tracker
  // software computes 1024-based sizes but often mislabels them.
  const factors: Record<string, number> = {
    b: 1 / (1024 ** 3), kib: 1 / (1024 ** 2), mib: 1 / 1024,
    gib: 1, tib: 1024, pib: 1048576,
    kb: 1 / (1024 ** 2), mb: 1 / 1024, gb: 1, tb: 1024, pb: 1048576,
  };
  return parseFloat(m[1]) * (factors[m[2].toLowerCase()] ?? 1);
}

type DurUnit = 'year' | 'month' | 'week' | 'day' | 'hour' | 'minute' | 'second' | null;

const DUR_SECONDS: Record<Exclude<DurUnit, null>, number> = {
  year: 365 * 86400, month: 30 * 86400, week: 7 * 86400,
  day: 86400, hour: 3600, minute: 60, second: 1,
};

/**
 * Map a unit token to its duration unit. Case matters where tracker layouts
 * rely on it: "M" is months against "m" minutes in English, and "S" is
 * semanas (weeks) against "s" seconds on Portuguese sites. A bare lowercase
 * "m" is ambiguous and gets resolved by position in parseSeedTime.
 */
function classifyDurUnit(raw: string): DurUnit {
  switch (raw) {
    case 'y': case 'Y': case 'a': case 'A':
    case 'ano': case 'anos': case 'year': case 'years':
      return 'year';
    case 'M': case 'mo': case 'mes': case 'meses': case 'month': case 'months':
      return 'month';
    case 'w': case 'W': case 'S':
    case 'sem': case 'semana': case 'semanas': case 'week': case 'weeks':
      return 'week';
    case 'd': case 'D': case 'dia': case 'dias': case 'day': case 'days':
      return 'day';
    case 'h': case 'H': case 'hora': case 'horas': case 'hour': case 'hours':
      return 'hour';
    case 'min': case 'mins':
    case 'minuto': case 'minutos': case 'minute': case 'minutes':
      return 'minute';
    case 's': case 'seg': case 'segs': case 'segundo': case 'segundos':
    case 'sec': case 'secs': case 'second': case 'seconds':
      return 'second';
  }
  return null;
}

/**
 * Parse seed time string → total seconds.
 * Splits the string into value/unit pairs so every occurrence of a unit
 * counts, and understands the Portuguese unit letters UNIT3D renders in
 * pt-BR ("178a 2m 4d 20h 52m 38s" = 178 years, 2 months … 52 minutes).
 * Also accepts a plain integer/float as raw seconds.
 * Mirrors parse.SeedTimeToSeconds on the Go side.
 */
export function parseSeedTime(str: unknown): number | null {
  if (str == null || str === '') return null;
  const s = String(str).trim();
  if (!s) return null;
  // Plain number → raw seconds
  if (/^\d+(\.\d+)?$/.test(s)) return Math.round(parseFloat(s));

  const tokens = [...s.matchAll(/(\d+(?:\.\d+)?)\s*([A-Za-zÀ-ÿ]+)/g)];
  if (!tokens.length) return null;

  const units = tokens.map((t) => classifyDurUnit(t[2]));
  // A bare lowercase "m" means minutes in English layouts and months in
  // Portuguese ones. UNIT3D renders units in descending order, so an "m" with
  // a day, week or hour after it occupies the month slot; later ones are
  // minutes.
  tokens.forEach((t, i) => {
    if (units[i] !== null || t[2] !== 'm') return;
    units[i] = 'minute';
    for (let j = i + 1; j < units.length; j++) {
      if (units[j] === 'day' || units[j] === 'week' || units[j] === 'hour') {
        units[i] = 'month';
        break;
      }
    }
  });

  let total = 0, found = false;
  tokens.forEach((t, i) => {
    const u = units[i];
    if (!u) return;
    found = true;
    total += parseFloat(t[1]) * DUR_SECONDS[u];
  });
  return found ? Math.round(total) : null;
}

/** Calculate account age in whole days from a YYYY-MM-DD string */
export function memberDays(d: string): number | null {
  if (!d) return null;
  const j = new Date(d);
  if (isNaN(j.getTime())) return null;
  return Math.floor((Date.now() - j.getTime()) / 86_400_000);
}

/** Format account age in days to "1Y 2M 3W 4D" */
export function memberDur(d: string): string {
  const days = memberDays(d);
  if (days === null || days < 0) return '—';
  if (days === 0) return '0D';
  const Y = Math.floor(days / 365), r1 = days % 365;
  const M = Math.floor(r1 / 30),   r2 = r1 % 30;
  const W = Math.floor(r2 / 7),    D = r2 % 7;
  return ([[Y,'Y'],[M,'M'],[W,'W'],[D,'D']] as [number,string][])
    .filter(([v]) => v)
    .map(([v, u]) => `${v}${u}`)
    .join(' ') || '0D';
}

/**
 * Parse account-age target: plain number (days) or "Y M W D" format.
 * Returns total days.
 */
export function parseAgeDays(str: string): number | null {
  if (!str) return null;
  const s = String(str).trim();
  if (/^\d+$/.test(s)) return parseInt(s);
  let total = 0;
  const yr = s.match(/(\d+)\s*Y/i); if (yr) total += parseInt(yr[1]) * 365;
  const mo = s.match(/(\d+)\s*M/);  if (mo) total += parseInt(mo[1]) * 30;
  const wk = s.match(/(\d+)\s*W/i); if (wk) total += parseInt(wk[1]) * 7;
  const dd = s.match(/(\d+)\s*D/i); if (dd) total += parseInt(dd[1]);
  return total > 0 ? total : null;
}

/** Build a favicon URL from a tracker URL */
export function getFaviconUrl(trackerUrl: string): string {
  try {
    const u = new URL(trackerUrl);
    return `${u.protocol}//${u.host}/favicon.ico`;
  } catch { return ''; }
}
